File: OpenFlix-tvOS/OpenFlix-tvOS/Models/Domain/Channel.swift
import Foundation

// MARK: - Channel

struct Channel: Identifiable, Hashable {
    let id: String
    let number: Int?
    let name: String
    let logo: String?
    let sourceId: String?
    let sourceName: String?
    let streamUrl: String?
    let enabled: Bool
    let isFavorite: Bool
    let group: String?
    let archiveEnabled: Bool
    let archiveDays: Int
    var nowPlaying: Program?
    var nextProgram: Program?

    // MARK: - Computed Properties

    var displayNumber: String {
        if let number = number {
            return "\(number)"
        }
        return ""
    }

    var displayName: String {
        if let number = number {
            return "\(number) - \(name)"
        }
        return name
    }

    var isHD: Bool {
        name.uppercased().contains("HD") ||
        name.uppercased().contains("FHD") ||
        name.uppercased().contains("4K") ||
        name.uppercased().contains("UHD")
    }

    var sortKey: Double {
        Double(number ?? Int.max)
    }

    // MARK: - Hashable

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: Channel, rhs: Channel) -> Bool {
        lhs.id == rhs.id
    }
}

// MARK: - DTO Mapping

extension ChannelDTO {
    func toDomain() -> Channel {
        Channel(
            id: safeId,
            number: number,
            name: safeName,
            logo: logo ?? thumb,
            sourceId: sourceId?.stringValue,
            sourceName: sourceName,
            streamUrl: streamUrl,
            enabled: enabled ?? true,
            isFavorite: isFavorite ?? false,
            group: group ?? category,
            archiveEnabled: archiveEnabled ?? false,
            archiveDays: archiveDays ?? 0,
            nowPlaying: nowPlaying?.toDomain(),
            nextProgram: nextProgram?.toDomain()
        )
    }
}

// MARK: - Channel Group

struct ChannelGroup: Identifiable {
    let id: Int
    let name: String
    let enabled: Bool
    let members: [ChannelGroupMember]
}

struct ChannelGroupMember: Identifiable {
    var id: String { channelId }
    let channelId: String
    let priority: Int
    let channelName: String?
}

extension ChannelGroupDTO {
    func toDomain() -> ChannelGroup {
        ChannelGroup(
            id: id,
            name: name,
            enabled: enabled ?? true,
            members: members?.map { ChannelGroupMember(channelId: $0.channelId, priority: $0.priority, channelName: $0.channelName) } ?? []
        )
    }
}

File: openflix-kmp/shared/src/commonMain/kotlin/com/openflix/domain/model/Channel.kt
package com.openflix.domain.model

data class Channel(
    val id: String,
    val number: Int? = null,
    val name: String,
    val logo: String? = null,
    val sourceId: String? = null,
    val sourceName: String? = null,
    val streamUrl: String? = null,
    val enabled: Boolean = true,
    val isFavorite: Boolean = false,
    val group: String? = null,
    val archiveEnabled: Boolean = false,
    val archiveDays: Int = 0,
    var nowPlaying: Program? = null,
    var nextProgram: Program? = null
) {
    val displayNumber: String get() = number?.toString() ?: ""

    val displayName: String get() = if (number != null) "$number - $name" else name

    val isHD: Boolean get() {
        val upper = name.uppercase()
        return upper.contains("HD") || upper.contains("FHD") ||
               upper.contains("4K") || upper.contains("UHD")
    }

    val sortKey: Double get() = number?.toDouble() ?: Double.MAX_VALUE
}

data class ChannelGroup(
    val id: Int,
    val name: String,
    val enabled: Boolean = true,
    val members: List<ChannelGroupMember> = emptyList()
)

data class ChannelGroupMember(
    val channelId: String,
    val priority: Int,
    val channelName: String? = null
)

data class ChannelWithPrograms(
    val channel: Channel,
    val programs: List<Program> = emptyList()
) {
    val id: String get() = channel.id

    fun programAt(timestamp: Long): Program? {
        return programs.firstOrNull { it.startTimeMs <= timestamp && it.endTimeMs > timestamp }
    }

    val currentProgram: Program? get() = programAt(currentTimeMs())

    val upcomingPrograms: List<Program> get() {
        val now = currentTimeMs()
        return programs.filter { it.startTimeMs > now }.sortedBy { it.startTimeMs }
    }
}

expect fun currentTimeMs(): Long
